File: src/y18/Day19.kt
package y18

private const val day = 19

private data class Line(val i: String, val a: Int, val b: Int, val c: Int)

@Suppress("UNUSED_PARAMETER")
private data class Registers19(val registers: MutableList<Int>) {
    fun set(register: Int, value: Int): Registers19 = registers
        .toMutableList()
        .also { it[register] = value }
        .let { Registers19(it) }

    fun get(register: Int) = registers[register]

    private fun bool(b: Boolean) = if (b) 1 else 0

    fun exec(s: String, a: Int, b: Int, c: Int) =
        when (s) {
            "addr" -> set(c, get(a) + get(b))
            "addi" -> set(c, get(a) + b)
            "mulr" -> set(c, get(a) * get(b))
            "muli" -> set(c, get(a) * b)
            "banr" -> set(c, get(a) and get(b))
            "bani" -> set(c, get(a) and b)
            "borr" -> set(c, get(a) or get(b))
            "bori" -> set(c, get(a) or b)
            "setr" -> set(c, get(a))
            "seti" -> set(c, a)
            "gtir" -> set(c, bool(a > get(b)))
            "gtri" -> set(c, bool(get(a) > b))
            "gtrr" -> set(c, bool(get(a) > get(b)))
            "eqir" -> set(c, bool(a == get(b)))
            "eqri" -> set(c, bool(get(a) == b))
            "eqrr" -> set(c, bool(get(a) == get(b)))
            else -> this
        }

    fun exec(line: Line) = exec(line.i, line.a, line.b, line.c)
}

fun main() {
    fun executeProgram(input: List<String>, init: Registers19): Int {
        var state = init
        val pointer = input.first().substringAfter("#ip ").toInt()
        val program = input.drop(1)
            .map { it.split(" ") }
            .map { (i, a, b, c) -> Line(i, a.toInt(), b.toInt(), c.toInt()) }

        var j = 0L
        while (true) {
            val idx = state.registers[pointer]
            if (idx !in program.indices) break
            val instruction = program[idx]
            state = state.exec(instruction)
            state.registers[pointer] = 1 + state.registers[pointer]

            if (++j % 12345678L == 476L)
                println("$j / $state")
        }
        return state.registers.first()
    }

    fun part1(input: List<String>): Int =
        executeProgram(input, Registers19(mutableListOf(0, 0, 0, 0, 0, 0)))

    @Suppress("UNUSED_VARIABLE")
    fun part2(): Long {
        // See Day19.js for a JS version of the instructions, where registries are named a,b,c,d,e,goto.

        // After the initial steps, the program enters a loop that can be summarized as
        //language=JavaScript
        val actualProgram = """
            do {
                b=1;
                do {
                    e = d*b;
                    if (e===c) a+=d;
                    b++;
                } while (b<=c);
                d++
            } while(d<=c);
        """.trimIndent()

        // This just adds to a (goal registry) all the divisors of value in registry c, which has value 10551277

        // so...
        val c = 10551277L
        return (1L..c).asSequence().filter { c % it == 0L }.sumOf { it }
    }

    val input = readInput("Day${day}")
    println(part1(input))
    println(part2())
}

File: src/y18/Day19.js
let j = 0
let a = 1
let b = 0
let c = 0
let d = 0
let e = 0
let goto = 0
const registers = () => ({a, b, c, d, e, goto})

const s0 = {} // all values ever held by r0 = a
const s2 = {} // all values ever held by r2 = c

function logAll(s) {
    console.log(j, ":", s, registers())
    console.log(s0)
    console.log(s2)
}

const program = [
    (/* 0 */) => goto = 16,
    (/* 1 */) => {
        d = 1;
        logAll("d = " + d)
    },
    (/* 2 */) => b = 1,
    (/* 3 */) => e = d * b,
    (/* 4 */) => {
        if (e === c) {
            e = 1;
            goto = 6
        } else {
            e = 0;
            goto = 7
        }
    },
    (/* 5 */) => goto = e + 5,
    (/* 6 */) => goto = 7,
    (/* 7 */) => {
        a = d + a;
        logAll("a = d + a = " + a)
        s0[a] = 1 + (s0[a] ?? 0)
    },
    (/* 8 */) => b = b + 1,
    (/* 9 */) => {
        if (b > c) {
            e = 1;
            goto=11
        } else {
            e = 0;
            goto = 2
        }
    },
    (/* 10 */) => goto = e + 10,
    (/* 11 */) => goto = 2,
    (/* 12 */) => {
        d = d + 1;
        logAll("d = d + 1 = " + d)
    },
    (/* 13 */) => {
        if (d <= c) {
            e = 0;
            goto = 1;
        } else {
            e = 1;
            goto = 9999999;
        }
    },
    (/* 14 */) => goto = e + 14,
    (/* 15 */) => goto = 1,
    (/* 16 */) => goto = 9999999,
    (/* 17 */) => {
        c = c + 2;
        logAll("c = c + 2 = " + c)
        s2[c] = 1 + (s2[c] ?? 0)
    },
    (/* 18 */) => {
        c = c * c;
        logAll("c = c * c = " + c)
        s2[c] = 1 + (s2[c] ?? 0)
    },
    (/* 19 */) => {
        c = c * 19;
        logAll("c = c * 19 = " + c)
        s2[c] = 1 + (s2[c] ?? 0)
    },
    (/* 20 */) => {
        c = c * 11;
        logAll("c = c * 11 = " + c)
        s2[c] = 1 + (s2[c] ?? 0)
    },
    (/* 21 */) => e = e + 1,
    (/* 22 */) => e = e * 22,
    (/* 23 */) => e = e + 19,
    (/* 24 */) => {
        c = c + e;
        logAll("c = c + e = " + c)
        s2[c] = 1 + (s2[c] ?? 0)
    },
    (/* 25 */) => goto = a + 25,
    (/* 26 */) => goto = 0,
    (/* 27 */) => e = 27,
    (/* 28 */) => e = e * 28,
    (/* 29 */) => e = e + 29,
    (/* 30 */) => e = e * 30,
    (/* 31 */) => e = e * 14,
    (/* 32 */) => e = e * 32,
    (/* 33 */) => {
        c = c + e;
        logAll("c = c + e = " + c)
        s2[c] = 1 + (s2[c] ?? 0)
    },
    (/* 34 */) => {
        a = 0;
        logAll("a = " + a)
        s0[a] = 1 + (s0[a] ?? 0)
    },
    (/* 35 */) => goto = 0,
]

do {
    j++
    program[goto]()
    goto++
} while (0 <= goto && goto < program.length);

console.log(registers())
